File: lib/commands/command.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'fileutils'
require 'optparse'

def ignore_parents(dir)
  Dir.entries(File.expand_path(dir)).reject{ |entry| entry == '.' || entry == '..' }
end

def read_tool_versions
  plugins = []
  File.open(File.expand_path("~/.tool-versions")) do |f|
    f.each do |line|
      plugin, *versions = line.split(/\s+/)
      plugins << [plugin, versions]
    end
  end
  plugins
end

def main
  options = {}
  OptionParser.new do |opts|
    opts.on('-f', '--[no-]force', 'Purge')
    opts.on('-h', '--help', 'Help') do |_h|
      puts opts
      exit
    end
  end.parse!(into: options)

  tuples = []
  read_tool_versions.each do |(plugin, versions)|
    versions.each do |version|
      tuples << [plugin, version]
    end
  end

  plugins = tuples.map(&:first).uniq

  ignore_parents("~/.asdf/installs").each do |plugin|
    preserve_plugin = plugins.include?(plugin)
    ignore_parents("~/.asdf/installs/#{plugin}").each do |version|
      next if preserve_plugin && tuples.include?([plugin, version])
      full_path = File.expand_path("~/.asdf/installs/#{plugin}/#{version}")
      FileUtils.remove_dir(full_path, true) if options[:force]
      puts "- #{full_path}"
    end
  end
end

main

File: lib/commands/command.py
#!/usr/bin/env python3

import os
import re
import shutil

from optparse import OptionParser
from pathlib import Path

def read_tool_versions():
    tuples = []
    with open(os.path.expanduser('~/.tool-versions')) as f:
        for line in f.readlines():
            splited = re.split('\s+', line.strip())
            for v in splited[1:]:
                tuples.append((splited[0], v))
    return tuples

def main():
    parser = OptionParser()
    parser.add_option('-f', '--force', help='Purge', action='store_true')
    (options, _args) = parser.parse_args()

    tuples = read_tool_versions()
    reserved_plugins = set(t[0] for t in tuples)
    for plugin_dir in Path(os.path.expanduser('~/.asdf/installs')).iterdir():
        plugin = os.path.basename(plugin_dir)
        if plugin not in reserved_plugins:
            print('- {}'.format(plugin_dir))
            if options.force:
                shutil.rmtree(plugin_dir)
        else:
            for version_dir in Path(plugin_dir).iterdir():
                version = os.path.basename(version_dir)
                if (plugin, version) not in tuples:
                    print('- {}'.format(version_dir))
                    if options.force:
                        shutil.rmtree(version_dir)


if __name__ == '__main__':
    main()
